File: app/wanes_app/lib/models/marketplace_models.dart
import 'models.dart';

/// One driver's offer on a request, as its riders compare them
/// (`GET ride-requests/{id}/offers`).
class RideOffer {
  const RideOffer({
    required this.interestId,
    required this.driverId,
    required this.driverName,
    required this.pricePerSeat,
    this.driverRating = 0,
    this.driverRatingCount = 0,
    this.driverTrips = 0,
    this.driverCompletionRate,
    this.driverVerified = false,
    this.vehicleLabel = '',
    this.vehicleColor = '',
    this.vehicleSeats = 0,
    this.seatsOffered,
    this.minPassengers,
    this.message,
    this.offeredAt,
  });

  final int interestId;
  final int driverId;
  final String driverName;
  final double driverRating;
  final int driverRatingCount;
  final int driverTrips;
  final double? driverCompletionRate;
  final bool driverVerified;
  final String vehicleLabel;
  final String vehicleColor;
  final int vehicleSeats;
  final double pricePerSeat;
  final int? seatsOffered;

  /// A conditional offer: the trip runs once this many seats are held.
  final int? minPassengers;
  final String? message;
  final DateTime? offeredAt;

  factory RideOffer.fromJson(Map<String, dynamic> j) => RideOffer(
        interestId: (j['interestId'] as num).toInt(),
        driverId: (j['driverId'] as num?)?.toInt() ?? 0,
        driverName: j['driverName'] as String? ?? '',
        driverRating: (j['driverRating'] as num?)?.toDouble() ?? 0,
        driverRatingCount: (j['driverRatingCount'] as num?)?.toInt() ?? 0,
        driverTrips: (j['driverTrips'] as num?)?.toInt() ?? 0,
        driverCompletionRate: (j['driverCompletionRate'] as num?)?.toDouble(),
        driverVerified: j['driverVerified'] as bool? ?? false,
        vehicleLabel: j['vehicleLabel'] as String? ?? '',
        vehicleColor: j['vehicleColor'] as String? ?? '',
        vehicleSeats: (j['vehicleSeats'] as num?)?.toInt() ?? 0,
        pricePerSeat: (j['pricePerSeat'] as num?)?.toDouble() ?? 0,
        seatsOffered: (j['seatsOffered'] as num?)?.toInt(),
        minPassengers: (j['minPassengers'] as num?)?.toInt(),
        message: j['message'] as String?,
        offeredAt: parseServerDate(j['offeredAt'] as String?),
      );
}

/// A driver's route alert, or a watch on one request (`me/demand-alerts`).
class DemandAlert {
  const DemandAlert({
    required this.id,
    required this.originAddress,
    required this.destinationAddress,
    required this.minSeats,
    this.radiusMeters = 3000,
    this.rideRequestId,
    this.recurringOnly = false,
    this.isActive = true,
    this.lastNotifiedAt,
    this.notifiedCount = 0,
  });

  final int id;
  final String originAddress;
  final String destinationAddress;
  final int radiusMeters;
  final int minSeats;
  final int? rideRequestId;

  /// A route alert that fires for recurring requests only.
  final bool recurringOnly;
  final bool isActive;
  final DateTime? lastNotifiedAt;
  final int notifiedCount;

  bool get isWatch => rideRequestId != null;

  factory DemandAlert.fromJson(Map<String, dynamic> j) => DemandAlert(
        id: (j['id'] as num).toInt(),
        originAddress: j['originAddress'] as String? ?? '',
        destinationAddress: j['destinationAddress'] as String? ?? '',
        radiusMeters: (j['radiusMeters'] as num?)?.toInt() ?? 3000,
        minSeats: (j['minSeats'] as num?)?.toInt() ?? 1,
        rideRequestId: (j['rideRequestId'] as num?)?.toInt(),
        recurringOnly: j['recurringOnly'] as bool? ?? false,
        isActive: j['isActive'] as bool? ?? true,
        lastNotifiedAt: parseServerDate(j['lastNotifiedAt'] as String?),
        notifiedCount: (j['notifiedCount'] as num?)?.toInt() ?? 0,
      );
}

/// Why a driver cancelled (`Shareds.Enums.CancelReason`).
enum CancelReason {
  personal(1, 'cancel.reasonPersonal'),
  vehicleProblem(2, 'cancel.reasonVehicle'),
  safetyConcern(3, 'cancel.reasonSafety'),
  emergency(4, 'cancel.reasonEmergency'),
  routeOrTimeChanged(5, 'cancel.reasonChanged'),
  other(9, 'cancel.reasonOther');

  const CancelReason(this.value, this.labelKey);

  final int value;
  final String labelKey;
}

/// How a cancellation is recorded (`Shareds.Enums.ReliabilityEventKind`).
enum ReliabilityKind {
  freeCancel(1, 'reliability.kindFree'),
  cancel(2, 'reliability.kindCancel'),
  lateCancel(3, 'reliability.kindLate'),
  riderLateCancel(4, 'reliability.kindRiderLate'),
  riderNoShow(5, 'reliability.kindNoShow'),
  seriesSkip(6, 'reliability.kindSeriesSkip'),
  seriesEnd(7, 'reliability.kindSeriesEnd');

  const ReliabilityKind(this.value, this.labelKey);

  final int value;
  final String labelKey;

  static ReliabilityKind fromValue(int? v) =>
      ReliabilityKind.values.firstWhere((k) => k.value == v, orElse: () => ReliabilityKind.freeCancel);
}

/// What cancelling a trip would cost the driver (`GET trips/{id}/cancel-preview`).
class CancelPreview {
  const CancelPreview({
    required this.kind,
    required this.points,
    required this.ridersAffected,
    required this.reasonRequired,
    this.pointsAfter = 0,
    this.warnPoints = 3,
    this.suspendPoints = 5,
    this.windowDays = 30,
    this.wouldSuspend = false,
    this.ridersRequeued = false,
  });

  final ReliabilityKind kind;
  final int points;
  final int ridersAffected;
  final bool reasonRequired;
  final int pointsAfter;
  final int warnPoints;
  final int suspendPoints;
  final int windowDays;
  final bool wouldSuspend;
  final bool ridersRequeued;

  factory CancelPreview.fromJson(Map<String, dynamic> j) => CancelPreview(
        kind: ReliabilityKind.fromValue((j['kind'] as num?)?.toInt()),
        points: (j['points'] as num?)?.toInt() ?? 0,
        ridersAffected: (j['ridersAffected'] as num?)?.toInt() ?? 0,
        reasonRequired: j['reasonRequired'] as bool? ?? false,
        pointsAfter: (j['pointsAfter'] as num?)?.toInt() ?? 0,
        warnPoints: (j['warnPoints'] as num?)?.toInt() ?? 3,
        suspendPoints: (j['suspendPoints'] as num?)?.toInt() ?? 5,
        windowDays: (j['windowDays'] as num?)?.toInt() ?? 30,
        wouldSuspend: j['wouldSuspend'] as bool? ?? false,
        ridersRequeued: j['ridersRequeued'] as bool? ?? false,
      );
}

/// A user's own reliability record (`GET me/reliability`).
class ReliabilityRecord {
  const ReliabilityRecord({
    this.pointsInWindow = 0,
    this.windowDays = 30,
    this.warnPoints = 3,
    this.suspendPoints = 5,
    this.suspendedUntil,
    this.tripsAsDriver = 0,
    this.driverCancellations = 0,
    this.completionRate,
    this.riderLateCancels = 0,
    this.riderNoShows = 0,
    this.recent = const [],
  });

  final int pointsInWindow;
  final int windowDays;
  final int warnPoints;
  final int suspendPoints;
  final DateTime? suspendedUntil;
  final int tripsAsDriver;
  final int driverCancellations;
  final double? completionRate;
  final int riderLateCancels;
  final int riderNoShows;
  final List<ReliabilityEntry> recent;

  bool get isSuspended => suspendedUntil != null && suspendedUntil!.isAfter(DateTime.now());

  factory ReliabilityRecord.fromJson(Map<String, dynamic> j) => ReliabilityRecord(
        pointsInWindow: (j['pointsInWindow'] as num?)?.toInt() ?? 0,
        windowDays: (j['windowDays'] as num?)?.toInt() ?? 30,
        warnPoints: (j['warnPoints'] as num?)?.toInt() ?? 3,
        suspendPoints: (j['suspendPoints'] as num?)?.toInt() ?? 5,
        suspendedUntil: parseServerDate(j['suspendedUntil'] as String?),
        tripsAsDriver: (j['tripsAsDriver'] as num?)?.toInt() ?? 0,
        driverCancellations: (j['driverCancellations'] as num?)?.toInt() ?? 0,
        completionRate: (j['completionRate'] as num?)?.toDouble(),
        riderLateCancels: (j['riderLateCancels'] as num?)?.toInt() ?? 0,
        riderNoShows: (j['riderNoShows'] as num?)?.toInt() ?? 0,
        recent: ((j['recent'] as List?) ?? const [])
            .map((e) => ReliabilityEntry.fromJson(e as Map<String, dynamic>))
            .toList(),
      );
}

class ReliabilityEntry {
  const ReliabilityEntry({
    required this.kind,
    required this.points,
    this.tripId,
    this.isWaived = false,
    this.needsReview = false,
    this.createdAt,
  });

  final ReliabilityKind kind;
  final int points;
  final int? tripId;
  final bool isWaived;
  final bool needsReview;
  final DateTime? createdAt;

  factory ReliabilityEntry.fromJson(Map<String, dynamic> j) => ReliabilityEntry(
        kind: ReliabilityKind.fromValue((j['kind'] as num?)?.toInt()),
        points: (j['points'] as num?)?.toInt() ?? 0,
        tripId: (j['tripId'] as num?)?.toInt(),
        isWaived: j['isWaived'] as bool? ?? false,
        needsReview: j['needsReview'] as bool? ?? false,
        createdAt: parseServerDate(j['createdAt'] as String?),
      );
}

/// The server's answer to an SOS or report.
class SafetyIncidentResult {
  const SafetyIncidentResult({
    required this.id,
    this.emergencyNumber = '911',
    this.emergencyContactNotified = false,
  });

  final int id;
  final String emergencyNumber;
  final bool emergencyContactNotified;

  factory SafetyIncidentResult.fromJson(Map<String, dynamic> j) => SafetyIncidentResult(
        id: (j['id'] as num).toInt(),
        emergencyNumber: (j['emergencyNumber'] as String?)?.trim().isNotEmpty == true
            ? (j['emergencyNumber'] as String).trim()
            : '911',
        emergencyContactNotified: j['emergencyContactNotified'] as bool? ?? false,
      );
}

/// A "follow my trip" link.
class ShareLink {
  const ShareLink({required this.token, this.url});

  final String token;

  /// The full link when the admin set a public address; null otherwise.
  final String? url;

  factory ShareLink.fromJson(Map<String, dynamic> j) =>
      ShareLink(token: j['token'] as String? ?? '', url: j['url'] as String?);
}

/// Versioned terms the user agrees to (`Shareds.Enums.AcknowledgementKind`).
enum AcknowledgementKindWire {
  riderSafety(1),
  driverSafety(2),
  riderSharedRide(3),
  driverSharedTrip(4);

  const AcknowledgementKindWire(this.value);
  final int value;

  static AcknowledgementKindWire? fromValue(int? v) {
    for (final k in values) {
      if (k.value == v) return k;
    }
    return null;
  }
}

File: backend/Wanes/Areas/Services/Marketplace/Models/MarketplaceModels.cs
using System.ComponentModel.DataAnnotations;
using Wanes.Areas.Domain.Marketplace;
using Wanes.Areas.Domain.RideRequests;
using Wanes.Areas.Domain.RiderTrips;
using Wanes.Shareds.Enums;
using Wanes.Shareds.Models;

namespace Wanes.Areas.Services.Marketplace.Models;

// ── Acknowledgements ─────────────────────────────────────────────────────────

public class AcknowledgementInput
{
    [EnumDataType(typeof(AcknowledgementKind))]
    public AcknowledgementKind Kind { get; set; }

    [Range(1, 1000)]
    public int Version { get; set; } = 1;
}

public class AcknowledgementOutput
{
    public AcknowledgementKind Kind { get; set; }
    public int Version { get; set; }
    public DateTime AcceptedAt { get; set; }

    public AcknowledgementOutput() { }

    public AcknowledgementOutput(UserAcknowledgement e)
    {
        Kind = e.Kind;
        Version = e.Version;
        AcceptedAt = e.AcceptedAt;
    }
}

// ── Demand alerts ────────────────────────────────────────────────────────────

/// <summary>
/// A route to watch, or — with <see cref="RideRequestId"/> — one request to
/// watch. For a request watch the route fields are ignored and copied from it.
/// </summary>
public class DemandAlertInput
{
    public GeoPoint? Origin { get; set; }
    public GeoPoint? Destination { get; set; }

    [Range(500, 50_000)]
    public int RadiusMeters { get; set; } = 3000;

    [Range(1, RiderTripRules.MaxSeats)]
    public int MinSeats { get; set; } = 3;

    public int? RideRequestId { get; set; }

    /// <summary>A route alert that fires only for recurring requests.</summary>
    public bool RecurringOnly { get; set; }
}

public class DemandAlertOutput
{
    public int Id { get; set; }
    public int DriverId { get; set; }
    public string? DriverName { get; set; }
    public string OriginAddress { get; set; } = string.Empty;
    public double OriginLat { get; set; }
    public double OriginLng { get; set; }
    public string DestinationAddress { get; set; } = string.Empty;
    public double DestinationLat { get; set; }
    public double DestinationLng { get; set; }
    public int RadiusMeters { get; set; }
    public int MinSeats { get; set; }
    public int? RideRequestId { get; set; }
    public bool RecurringOnly { get; set; }
    public bool IsActive { get; set; }
    public DateTime? LastNotifiedAt { get; set; }
    public int NotifiedCount { get; set; }
    public DateTime CreatedAt { get; set; }

    public DemandAlertOutput() { }

    public DemandAlertOutput(DemandAlert a)
    {
        Id = a.Id;
        DriverId = a.DriverId;
        DriverName = a.Driver == null ? null : $"{a.Driver.FirstName} {a.Driver.LastName}".Trim();
        OriginAddress = a.OriginAddress;
        OriginLat = a.Origin.Y;
        OriginLng = a.Origin.X;
        DestinationAddress = a.DestinationAddress;
        DestinationLat = a.Destination.Y;
        DestinationLng = a.Destination.X;
        RadiusMeters = a.RadiusMeters;
        MinSeats = a.MinSeats;
        RideRequestId = a.RideRequestId;
        RecurringOnly = a.RecurringOnly;
        IsActive = a.IsActive;
        LastNotifiedAt = a.LastNotifiedAt;
        NotifiedCount = a.NotifiedCount;
        CreatedAt = a.CreationDate;
    }
}

// ── Reliability ──────────────────────────────────────────────────────────────

public class CancelTripInput
{
    [EnumDataType(typeof(CancelReason))]
    public CancelReason? Reason { get; set; }

    [StringLength(500)]
    public string? Note { get; set; }
}

/// <summary>What cancelling would cost, shown before the driver confirms.</summary>
public class CancelPreviewOutput
{
    public ReliabilityEventKind Kind { get; set; }
    public int Points { get; set; }
    public int RidersAffected { get; set; }
    public bool ReasonRequired { get; set; }

    /// <summary>Points in the current window after this one, against the thresholds.</summary>
    public int PointsAfter { get; set; }
    public int WarnPoints { get; set; }
    public int SuspendPoints { get; set; }
    public int WindowDays { get; set; }
    public bool WouldSuspend { get; set; }

    /// <summary>The riders will be put back on the market in a new request.</summary>
    public bool RidersRequeued { get; set; }
}

public class ReliabilityEventRow
{
    public int Id { get; set; }
    public int UserId { get; set; }
    public string? UserName { get; set; }
    public string? UserPhone { get; set; }
    public ActiveRole Role { get; set; }
    public ReliabilityEventKind Kind { get; set; }
    public int Points { get; set; }
    public int? TripId { get; set; }
    public int? BookingId { get; set; }
    public CancelReason? Reason { get; set; }
    public string? Note { get; set; }
    public int RidersAffected { get; set; }
    public int MinutesBeforeDeparture { get; set; }
    public bool NeedsReview { get; set; }
    public bool IsWaived { get; set; }
    public DateTime? WaivedAt { get; set; }
    public string? WaiveNote { get; set; }
    public DateTime CreatedAt { get; set; }

    public ReliabilityEventRow() { }

    public ReliabilityEventRow(ReliabilityEvent e)
    {
        Id = e.Id;
        UserId = e.UserId;
        UserName = e.User == null ? null : $"{e.User.FirstName} {e.User.LastName}".Trim();
        UserPhone = e.User?.Phone;
        Role = e.Role;
        Kind = e.Kind;
        Points = e.Points;
        TripId = e.TripId;
        BookingId = e.BookingId;
        Reason = e.Reason;
        Note = e.Note;
        RidersAffected = e.RidersAffected;
        MinutesBeforeDeparture = e.MinutesBeforeDeparture;
        NeedsReview = e.NeedsReview;
        IsWaived = e.IsWaived;
        WaivedAt = e.WaivedAt;
        WaiveNote = e.WaiveNote;
        CreatedAt = e.CreationDate;
    }
}

/// <summary>A user's own record: what counts now, and what it means.</summary>
public class ReliabilityOutput
{
    public int PointsInWindow { get; set; }
    public int WindowDays { get; set; }
    public int WarnPoints { get; set; }
    public int SuspendPoints { get; set; }
    public DateTime? SuspendedUntil { get; set; }
    public int TripsAsDriver { get; set; }
    public int DriverCancellations { get; set; }
    public double? CompletionRate { get; set; }
    public int RiderLateCancels { get; set; }
    public int RiderNoShows { get; set; }
    public List<ReliabilityEventRow> Recent { get; set; } = [];
}

public class WaiveInput
{
    [StringLength(500)]
    public string? Note { get; set; }

    /// <summary>The same note, under the name the console's edit form uses.</summary>
    [StringLength(500)]
    public string? WaiveNote { get; set; }
}

// ── Offers (riders choosing) ─────────────────────────────────────────────────

/// <summary>One driver's offer on a request, as its riders compare them.</summary>
public class OfferRow
{
    public int InterestId { get; set; }
    public int DriverId { get; set; }
    public string DriverName { get; set; } = string.Empty;
    public double DriverRating { get; set; }
    public int DriverRatingCount { get; set; }
    public int DriverTrips { get; set; }
    public double? DriverCompletionRate { get; set; }
    public bool DriverVerified { get; set; }
    public string? VehicleLabel { get; set; }
    public string? VehicleColor { get; set; }
    public int VehicleSeats { get; set; }
    public decimal PricePerSeat { get; set; }
    public int? SeatsOffered { get; set; }
    public int? MinPassengers { get; set; }
    public string? Message { get; set; }
    public DateTime OfferedAt { get; set; }

    public OfferRow() { }

    public OfferRow(DriverInterest i)
    {
        InterestId = i.Id;
        DriverId = i.DriverId;
        DriverName = i.Driver?.DisplayName ?? i.Driver?.FirstName ?? string.Empty;
        DriverRating = i.Driver?.RatingAvg ?? 0;
        DriverRatingCount = i.Driver?.RatingCount ?? 0;
        DriverTrips = i.Driver?.TripsAsDriver ?? 0;
        DriverCompletionRate = i.Driver == null
            ? null
            : ReliabilityRules.CompletionRate(i.Driver.TripsAsDriver, i.Driver.DriverCancellations);
        DriverVerified = i.Driver?.DriverStatus == DriverStatus.Verified;
        if (i.Vehicle != null)
        {
            VehicleLabel = $"{i.Vehicle.Make} {i.Vehicle.Model}".Trim();
            VehicleColor = i.Vehicle.Color;
            VehicleSeats = i.Vehicle.SeatCapacity;
        }
        PricePerSeat = i.PricePerSeat;
        SeatsOffered = i.SeatsOffered;
        MinPassengers = i.MinPassengers;
        Message = i.Message;
        OfferedAt = i.CreationDate;
    }
}

// ── Admin: ride requests ─────────────────────────────────────────────────────

public class RideRequestAdminRow
{
    public int Id { get; set; }
    public string OriginAddress { get; set; } = string.Empty;
    public string DestinationAddress { get; set; } = string.Empty;
    public DateTime DepartAt { get; set; }
    public int SeatsRequested { get; set; }
    public int RiderCount { get; set; }
    public string? AuthorName { get; set; }
    public RideRequestStatus Status { get; set; }
    public GenderPolicy DriverGenderPolicy { get; set; }
    public GenderPolicy CoRiderGenderPolicy { get; set; }
    public int InterestCount { get; set; }
    public DateTime? FirstInterestAt { get; set; }
    public DateTime? DecideAt { get; set; }
    public DateTime? NotifiedAt { get; set; }
    public int? MatchedTripId { get; set; }
    public int? ReopenedFromRequestId { get; set; }
    public int? ScheduleId { get; set; }
    public DateTime CreatedAt { get; set; }
}
